### PR TITLE
✨ (slope) redesign / TAS-722

### PR DESCRIPTION
**Pull request by sophiamersmann**
### Summary

- Entity names are usually only shown only on the right side
- If it would be otherwise difficult to match a label on the left to the corresponding slope, then the entity name is repeated
    - The heuristic here uses the number of levels used by the line legend, the actual threshold (4) is arbitrary of course
- We prefer to add labels on the left if they're also labelled on the right (this is only important if labels are dropped due to space constraints)
- Entity annotations are hidden on smaller screens
- Entity annotations are also shown in the tooltip (line charts show entity annotations in the tooltip as well) 

### Future improvements

- Show hidden line legend labels on hover
- Toggling the No Data section is quite disruptive. Once we make changes to the entity selector that account for the No data problem, we should remove the section

### Testing

The Multiembedder problem is hot-fixed on [this staging site](http://staging-site-slope-charts) for testing.

### CODE REVIEW

File: packages/@ourworldindata/components/src/TextWrap/TextWrap.tsx [hunk 1]
```diff
@@ -194,16 +205,37 @@ export class TextWrap {
...
+
     @computed get height(): number {
-        const { lines, lineHeight, fontSize } = this
-        if (lines.length === 0) return 0
-        return lines.length * lineHeight * fontSize
+        if (this.lineCount === 0) return 0
+        return this.lineCount * this.singleLineHeight
+    }
+
+    @computed get heightWithoutOffsetedLine(): number {
```
Comment (marcelgerber) [inline packages/@ourworldindata/components/src/TextWrap/TextWrap.tsx:221]: [The past form of offset is also offset](https://dictionary.cambridge.org/dictionary/english/offset) :)
```suggestion
    @computed get heightWithoutOffsetLine(): number {
```

File: packages/@ourworldindata/grapher/src/slopeCharts/SlopeChart.tsx [hunk 2]
```diff
@@ -493,43 +511,163 @@ export class SlopeChart
...
+    }
+
+    @computed get visibleLineLegendLabelsRight(): Set<SeriesName> {
+        return new Set(this.lineLegendRight?.visibleSeriesNames ?? [])
+    }
+
+    @computed get seriesSortedByImportanceForLineLegendLeft(): SeriesName[] {
+        return this.series
+            .map((s) => s.seriesName)
+            .toSorted((s1: SeriesName, s2: SeriesName): number => {
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/slopeCharts/SlopeChart.tsx:621]: While I love `toSorted`, and we actually do already polyfill it for older browsers, [it is not supported in Node 18.16](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSorted#browser_compatibility), which we currently use :(

**Review by marcelgerber — CHANGES_REQUESTED** [review]
Nice! I like that you could adapt the line legend code for your use case, and I very much like the resulting new slope style that comes out of all of that.

I'm sorry to say that I found a few critical flaws in the TextWrapGroup logic...

File: packages/@ourworldindata/grapher/src/slopeCharts/SlopeChart.tsx [hunk 3]
```diff
@@ -493,43 +511,163 @@ export class SlopeChart
...
+        return {
+            xAnchor: "start",
+            fontWeight: 700,
+        }
+    }
+
+    @computed private get lineLegendPropsLeft(): Partial<LineLegendProps> {
+        return {
+            xAnchor: "end",
+            fontWeight: this.showSeriesNamesInLineLegendLeft ? 700 : undefined,
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/slopeCharts/SlopeChart.tsx:561]: I don't think you need this, since it's part of the line legend code anyhow, right?

**Comment by danyx23** [comment]
Hey @marcelgerber, just a drive by comment to say it's very much appreciated that you give such in-depth feedback and think things through independently. Complex layout logic is certainly one of those areas where another pair of eyes is especially valuable as it's easy to miss edge cases. I know PR reviews can seem a bit tedious and "distracting from your cycle work", but it's very valuable that you take this time!

@sophiamersmann I hope you don't feel too bad about having to open this box again 😅. Thanks both of you for your diligent work!

File: packages/@ourworldindata/components/src/TextWrap/TextWrapGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+import { TextWrap } from "./TextWrap"
+import { splitIntoFragments } from "./TextWrapUtils"
+import { Bounds, last, max } from "@ourworldindata/utils"
+
+interface TextWrapFragment {
+    text: string
+    fontWeight?: number
+    // "always" places the fragment in a new line in all cases
+    // "avoid-wrap" places the fragment in a new line only if the fragment would wrap otherwise
+    newLine?: "always" | "avoid-wrap"
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply packages/@ourworldindata/components/src/TextWrap/TextWrapGroup.tsx:12]: I don't think `newLineBefore` would make it much clearer, but I added in a comment that this doesn't apply to the first fragment.

File: packages/@ourworldindata/components/src/TextWrap/TextWrapGroup.tsx [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                    textWrap.firstLineOffset > 0 &&
+                    isFirstTextWrapLine &&
+                    lastLine
+                ) {
+                    // if the current line is offsetted, add it to the previous line
+                    lastLine.fragments.push(fragment)
+                } else {
+                    // else, push a new line
+                    lines.push({
+                        textWrap,
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply packages/@ourworldindata/components/src/TextWrap/TextWrapGroup.tsx:181]: yeah, I just needed it for positioning, so it didn't really matter which one I used – by I refactored it now

File: packages/@ourworldindata/grapher/src/slopeCharts/SlopeChart.tsx (hunk 3 above)
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/slopeCharts/SlopeChart.tsx:561]: It isn't currently but I can make it so :)

**Comment by sophiamersmann** [comment]
I think this should fix it. If I understand correctly then it was one edge case (long second fragment) that was missed and then cascaded through the code. I also refactored the code a bit and added a few more test cases. The gist is in this [commit](https://github.com/owid/owid-grapher/pull/4222/commits/c272024a4109e70e34e84dbc2173230779cc8ee0). The rest of the PRs are some more style changes in response to Marwa's feedback.

**Comment by sophiamersmann** [comment]
### Merge activity

* **Dec 11, 7:47 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/4222).
